<HTML>
<HEAD>
<TITLE>Haplotype.pm</TITLE>
<LINK REV="made" HREF="mailto:smathias1@qwest.net">
</HEAD>

<BODY>

<A NAME="__index__"></A>
<!-- INDEX BEGIN -->

<UL>

	<LI><A HREF="#name">NAME</A></LI>
	<LI><A HREF="#synopsis">SYNOPSIS</A></LI>
	<LI><A HREF="#description">DESCRIPTION</A></LI>
	<LI><A HREF="#license">LICENSE</A></LI>
	<LI><A HREF="#feedback">FEEDBACK</A></LI>
	<LI><A HREF="#author  steve mathias">AUTHOR - Steve Mathias</A></LI>
	<LI><A HREF="#details">DETAILS</A></LI>
	<LI><A HREF="#imported packages">Imported Packages</A></LI>
	<LI><A HREF="#public methods">Public methods</A></LI>
	<UL>

		<LI><A HREF="#printxml">printXML</A></LI>
	</UL>

</UL>
<!-- INDEX END -->

<HR>
<P>
<H1><A NAME="name">NAME</A></H1>
<P>Genetics::Haplotype</P>
<P>
<HR>
<H1><A NAME="synopsis">SYNOPSIS</A></H1>
<PRE>

  $ht = new Genetics::Haplotype(name =&gt; '12pEA1.1',
                                importID =&gt; 268,
                                dateCreated =&gt; $today,
                                Keywords =&gt; [ {name =&gt; &quot;Test Data&quot;, 
                                               dataType =&gt; &quot;Boolean&quot;, 
                                               value =&gt; 1}, 
                                            ], 
                                MarkerCollection =&gt; {name =&gt; &quot;12pEA1&quot;, importID =&gt; 267},
                                Alleles =&gt; [ {name =&gt; 2, type =&gt; &quot;code&quot;}, 
                                             {name =&gt; &quot;C&quot;, type =&gt; &quot;nucleotide&quot;} ], 
                               ) ;</PRE>
<P>See the GenPerl Tutorial for more information.</P>
<P>
<HR>
<H1><A NAME="description">DESCRIPTION</A></H1>
<P>Haplotype objects represent a specific combination of alleles that are known, or 
predicted, to be segregating togehter.  The alleles in a haplotype may be experimentally 
determined or predicted statistically.</P>
<P>
<HR>
<H1><A NAME="license">LICENSE</A></H1>
<P>This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.</P>
<P>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</P>
<P>You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA</P>
<P>
<HR>
<H1><A NAME="feedback">FEEDBACK</A></H1>
<P>Currently, all feedback should be sent directly to the author.</P>
<P>
<HR>
<H1><A NAME="author  steve mathias">AUTHOR - Steve Mathias</A></H1>
<P>Email: <A HREF="mailto:mathias@genomica.com">mathias@genomica.com</A></P>
<P>Phone: (720) 565-4029</P>
<P>Address: Genomica Corporation 
         1745 38th Street
         Boulder, CO 80301</P>
<P>
<HR>
<H1><A NAME="details">DETAILS</A></H1>
<P>The rest of the documentation describes each of the object variables and 
methods. The names of internal variables and methods are preceded with an
underscore (_).</P>
<P>
<HR>
<H1><A NAME="imported packages">Imported Packages</A></H1>
<PRE>
 Genetics::Object       Superclass
 strict                 Just to be anal
 vars                   Global variables</PRE>
<P>
<HR>
<H1><A NAME="public methods">Public methods</A></H1>
<P>
<H2><A NAME="printxml">printXML</A></H2>
<PRE>
  Function  : Print an XML representation of the object.
  Argument  : A Genetics::StudyVariable object and the XML::Writer object being 
              used to generate the XML.
  Returns   : String
  Scope     : Public Instance Method
  Comments  : Calls Genetics::Object-&gt;printGeneralXML to generate XML elements
              common to all Genetics::Object objects.</PRE>

</BODY>

</HTML>
